packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,993 @@
+  ///
+  /// At most one [Router] can opt in to route name reporting. In common use
+  /// case, only the top-most [Router] created by [WidgetsApp.router] should
+  /// opt for route name reporting.
+  T get currentConfiguration => null;
+
+  /// Called by the [Router] to obtain the widget tree that represents the
+  /// current state.
+  ///
+  /// This is called whenever the [init] method's future completes, the
+  /// [pushRoute] method's future completes with the value true, the [popRoute]
+  /// method's future completes with the value true, or this object notifies its
+  /// clients (see the [Listenable] interface, which this interface includes).
+  /// In addition, it may be called at other times. It is important, therefore,
+  /// that the methods above do not update the state that the [build] method
+  /// uses before they complete their respective futures.
+  ///
+  /// Typically this method returns a suitably-configured [Navigator]. If you do
+  /// plan to create a navigator, consider using the
+  /// [PopNavigatorRouterDelegateMixin].
+  ///
+  /// This method must not return null.
+  ///
+  /// The `context` is the [Router]'s build context.
+  Widget build(BuildContext context);
+}
+
+/// A route name provider that provides route name for the [Router] widget
+///
+/// This provider is responsible for handing the route name through [value]
+/// getter and notifies listeners, typically the [Router] widget, when a new
+/// route name is available.
+///
+/// If you decide to opt for the route name reporting (by overrides the
+/// [RouterDelegate.currentConfiguration] to return non-null), overrides the
+/// [routerReportsNewRouteName] method to reports the route name back to
+/// the engine or uses the [RouteNameProviderReportBack] mixin to do so.
+///
+/// See also:
+///  * [PlatformRouteNameProvider], which wires up the provider with the
+///  [WidgetsBindingObserver.didPushRoute] to propagate platform push route
+///  intent to the [Router] widget.
+///  * [RouteNameProviderReportBack], which provides the implementation of
+///  [routerReportsNewRouteName].
+abstract class RouteNameProvider extends ValueListenable<String> {
+  /// A callback called when the [Router] widget has a new route name that is
+  /// potentially different from the existing route name in this provider.
+  ///
+  /// It is possible the new route name is the same as the current value in
+  /// this provider and no update is needed. The subclasses are responsible to
+  /// handle such cases.
+  void routerReportsNewRouteName(String routeName) { }
+}
+
+/// The mixin that implements the hook to report route name back to the
+/// engine.
+///
+/// Uses this mixin if you want to report the route name back to the engine.
+/// This is essential for the web application. The reported route names are
+/// used for constructing the browser history so that it can support the forward
+/// and backward buttons.
+mixin RouteNameProviderReportBack on RouteNameProvider {
```

Comment:
It's not clear to me what value this mixin adds. It's only used by one class, can that class just inhouse this functionality?